### ISSUE TITLE
alpha weight for negative loss

### ISSUE THREAD

**Issue by farisazhari** [artemmavrin/focal-loss#8]
Hi,

On line 512 and forward of _binary_focal_loss.py: 

    def _binary_focal_loss_from_probs(labels, p, gamma, pos_weight,
                                  label_smoothing):
    """Compute focal loss from probabilities.

    Parameters
    ----------
    labels : tensor-like
        Tensor of 0's and 1's: binary class labels.

    p : tf.Tensor
        Estimated probabilities for the positive class.

    gamma : float
        Focusing parameter.

    pos_weight : float or None
        If not None, losses for the positive class will be scaled by this
        weight.

    label_smoothing : float or None
        Float in [0, 1]. When 0, no smoothing occurs. When positive, the binary
        ground truth labels `y_true` are squeezed toward 0.5, with larger values
        of `label_smoothing` leading to label values closer to 0.5.

    Returns
    -------
    tf.Tensor
        The loss for each example.
    """
    # Predicted probabilities for the negative class
    q = 1 - p

    # For numerical stability (so we don't inadvertently take the log of 0)
    p = tf.math.maximum(p, _EPSILON)
    q = tf.math.maximum(q, _EPSILON)

    # Loss for the positive examples
    pos_loss = -(q ** gamma) * tf.math.log(p)
    if pos_weight is not None:
        pos_loss *= pos_weight

    # Loss for the negative examples
    neg_loss = -(p ** gamma) * tf.math.log(q)

    # Combine loss terms
    if label_smoothing is None:
        labels = tf.dtypes.cast(labels, dtype=tf.bool)
        loss = tf.where(labels, pos_loss, neg_loss)
    else:
        labels = _process_labels(labels=labels, label_smoothing=label_smoothing,
                                 dtype=p.dtype)
        loss = labels * pos_loss + (1 - labels) * neg_loss

    return loss

With the negative loss calculation:

    # Loss for the negative examples
    neg_loss = -(p ** gamma) * tf.math.log(q)

Shouldn't there be a '1-pos_weight' weighting factor just like for the pos_loss in the lines above it?

**Comment by artemmavrin** [comment]
Thanks for the question! The reason the negative component of the loss isn't scaled by `1-pos_weight` is that optimizing a weighted sum of the form `pA+(1-p)B`, where `0<p<1` is equivalent (up to a multiplicative constant) to optimizing a weighted sum `qA+B`, where `q>0`. 

This is because `pA+(1-p)B` can be written as `(1-p)((p/(1-p))A+B)=C(qA+B)`, where `C=1-p` (constant) and `q=p/(1-p)`. Likewise, `qA+B` can be written as `(q+1)((q/(q+1))A+(1/(q+1))B)=C(pA+(1-p)B)`, where `C=q+1` (constant) and `p=q/(q+1)`.

In `binary_focal_loss`, the weighted sum has the second form (`qA+B`, where `A=positive loss`, `B=negative loss`, and `q=pos_weight`), and `pos_weight` isn't restricted to being between 0 and 1. This design choice was made to be consistent with [`tf.nn.weighted_cross_entropy_from_logits`](https://www.tensorflow.org/api_docs/python/tf/nn/weighted_cross_entropy_with_logits).

**Comment by farisazhari** [comment]
Make sense now.

So let say based on the paper, for alpha=0.75, to get the same weighting factor, pos_weight should be 3. And for alpha=0.25, pos_weight should be 1/3. Obviously, the loss value will not be the same but at least one should be a factor or a multiple of the other. 

I'm I right?

**Comment by artemmavrin** [comment]
Yes, exactly! 😀

**Comment by farisazhari** [comment]
Thank You. Cheers
